cmd/osv-scanner/fix/noninteractive.go

```diff
@@ -0,0 +1,207 @@
+
+		if _, seen := seenVKs[vk]; maxUpgrades != 0 && !seen {
+			seenVKs[vk] = struct{}{}
+			patches = append(patches, p.DependencyPatch)
+			maxUpgrades--
+			numFixed += len(p.ResolvedVulns)
+		}
+
+		for _, rv := range p.ResolvedVulns {
+			uniqueVulns[vulnKey{id: rv.Vulnerability.ID, vk: vk}] = struct{}{}
+		}
+	}
+
+	return patches, numFixed, len(uniqueVulns) - numFixed
+}
+
+func autoRelock(ctx context.Context, r reporter.Reporter, opts osvFixOptions, maxUpgrades int) error {
+	r.Infof("Resolving %s...\n", opts.Manifest)
+	f, err := lockfile.OpenLocalDepFile(opts.Manifest)
+	if err != nil {
+		return err
+	}
+
+	manif, err := opts.ManifestRW.Read(f)
+	f.Close()
+	if err != nil {
+		return err
+	}
+
+	opts.Client.PreFetch(ctx, manif.Requirements, manif.FilePath)
+	res, err := resolution.Resolve(ctx, opts.Client, manif)
+	if err != nil {
+		return err
+	}
+
+	if errs := res.Errors(); len(errs) > 0 {
+		r.Warnf("WARNING: encountered %d errors during dependency resolution:\n", len(errs))
+		r.Warnf(resolutionErrorString(res, errs))
+	}
+
+	res.FilterVulns(opts.MatchVuln)
+	// TODO: count vulnerabilities per unique version as scan action does
+	totalVulns := len(res.Vulns)
+	r.Infof("Found %d vulnerabilities matching the filter\n", totalVulns)
+
+	allPatches, err := remediation.ComputeRelaxPatches(ctx, opts.Client, res, opts.RemediationOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := opts.Client.WriteCache(manif.FilePath); err != nil {
+		r.Warnf("WARNING: failed to write resolution cache: %v\n", err)
+	}
+
+	if len(allPatches) == 0 {
+		r.Infof("No dependency patches are possible\n")
+		r.Infof("REMAINING-VULNS: %d\n", totalVulns)
+		r.Infof("UNFIXABLE-VULNS: %d\n", totalVulns)
+
+		return nil
+	}
+
+	depPatches, nFixed, nUnfixable := autoChooseRelockPatches(allPatches, maxUpgrades)
+	r.Infof("Can fix %d/%d matching vulnerabilities by changing %d dependencies\n", nFixed, totalVulns, len(depPatches))
+	for _, p := range depPatches {
+		r.Infof("UPGRADED-PACKAGE: %s,%s,%s\n", p.Pkg.Name, p.OrigRequire, p.NewRequire)
+	}
+	r.Infof("REMAINING-VULNS: %d\n", totalVulns-nFixed)
+	r.Infof("UNFIXABLE-VULNS: %d\n", nUnfixable)
+
+	r.Infof("Rewriting %s...\n", opts.Manifest)
+	if err := manifest.Overwrite(opts.ManifestRW, opts.Manifest, manifest.ManifestPatch{Manifest: &manif, Deps: depPatches}); err != nil {
+		return err
+	}
+
+	if opts.Lockfile != "" || opts.RelockCmd != "" {
+		// We only recreate the lockfile if we know a lockfile already exists
+		// or we've been given a command to run.
+		r.Infof("Shelling out to regenerate lockfile...\n")
+		return regenerateLockfile(r, opts)
+	}
+
+	return nil
+}
+
+// returns the top {maxUpgrades} compatible patches, the number of vulns fixed, and the number unfixable vulns
+// if maxUpgrades is < 0, do as many patches as possible
+func autoChooseRelockPatches(diffs []resolution.ResolutionDiff, maxUpgrades int) ([]manifest.DependencyPatch, int, int) {
+	unfixableVulnIDs := make(map[string]struct{})
```

Comment:
add some high level comments explaining the logic in this function.

Reply:
Added